flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/data/RowDataNestProjection.java

```diff
@@ -0,0 +1,281 @@
+                                Types.StructType schema,
+                                Types.StructType rowStruct,
+                                Types.StructType projectType,
+                                int[][] projectedFields,
+                                boolean nestFlat) {
+    this.projectedFields = projectedFields;
+    this.nestFlat = nestFlat;
+
+    this.getters = new FieldGetter[projectedFields.length];
+    for (int i = 0; i < getters.length; i++) {
+      int[] projectedFieldOne = projectedFields[i];
+      int fieldIndex = projectedFieldOne[0];
+
+      Types.NestedField projectField = rowStruct.fields().get(fieldIndex);
+      Types.NestedField rowField = schema.field(projectField.fieldId());
+
+      getters[i] = createFieldGetter(rowType, rowField, fieldIndex, projectType, projectField, i, projectedFields,
+          projectedFieldOne);
+    }
+  }
+
+  private RowDataNestProjection(RowType rowType,
+                                Types.StructType schema,
+                                Types.StructType projectType,
+                                int[][] projectedFields,
+                                boolean nestFlat) {
+    this.projectedFields = projectedFields;
+    this.nestFlat = nestFlat;
+
+    this.getters = new FieldGetter[projectType.fields().size()];
+    for (int i = 0; i < getters.length; i++) {
+      int[] projectedFieldOne = projectedFields[0];
+      int fieldIndex = projectedFieldOne[0];
+      if (i == fieldIndex) {
+        Types.NestedField projectField = projectType.fields().get(i);
+        Types.NestedField rowField = schema.field(projectField.fieldId());
+
+        getters[i] = createFieldGetter(rowType, rowField, i, projectType, projectField, i, projectedFields,
+            projectedFieldOne);
+      }
+    }
+  }
+
+  private static FieldGetter createFieldGetter(RowType rowType,
+                                               Types.NestedField rowField,
+                                               int rowTypePosition,
+                                               Types.StructType projectFieldsType,
+                                               Types.NestedField projectField,
+                                               int projectFieldPosition,
+                                               int[][] projectedFields,
+                                               int[] projectedFieldOne) {
+    switch (projectField.type().typeId()) {
+      case STRUCT:
+        if (projectedFields == null || projectedFields[projectFieldPosition].length <= 1) {
+          return RowData.createFieldGetter(rowType.getTypeAt(rowTypePosition),
+              projectFieldsType.fields().indexOf(projectField));
+        }
+        RowType nestedRowType = (RowType) rowType.getTypeAt(rowTypePosition);
+        return row -> {
+          int[] target = new int[projectedFieldOne.length - 1];
+          System.arraycopy(projectedFieldOne, 1, target, 0, target.length);
```

Comment:
If we want to access the nested field from the original row, when we need to copy the project field index each time ?  That does not sounds acceptable for me,   could we choose another approach to track the nested project field index without cloning the index array ?

Reply:
Thanks, it does seem to bottleneck. I've put it outside the inner class so that it will only run during the planning phase, greatly reducing the number of replications.
For example: three times for `SELECT id, st.a, st.b, st2.a FROM t`, twice for `SELECT id, st.a, st.b FROM t`